lib/Default/MySqlProperties.class.php

```diff
@@ -0,0 +1,82 @@
+<?php declare(strict_types=1);
+
+use Dotenv\Dotenv;
+
+class MySqlProperties {
+	private const CONFIG_MYSQL_HOST = "MYSQL_HOST";
+	private const CONFIG_MYSQL_USER = "MYSQL_USER";
+	private const CONFIG_MYSQL_PASSWORD = "MYSQL_PASSWORD";
+	private const CONFIG_MYSQL_PORT = "MYSQL_PORT";
+	private const CONFIG_MYSQL_DATABASE = "MYSQL_DATABASE";
+	private string $host;
+	private string $user;
+	private string $password;
+	private int $port;
+	private string $databaseName;
+
+	public function __construct() {
+		$environmentFile = self::determineConfigEnvironmentFile();
+		$config = Dotenv::createArrayBacked(ROOT, $environmentFile);
+		$array = $config->load();
+		self::validateConfig($config);
+		[
+			MySqlProperties::CONFIG_MYSQL_HOST => $this->host,
+			MySqlProperties::CONFIG_MYSQL_USER => $this->user,
+			MySqlProperties::CONFIG_MYSQL_PASSWORD => $this->password,
+			MySqlProperties::CONFIG_MYSQL_PORT => $port,
+			MySqlProperties::CONFIG_MYSQL_DATABASE => $this->databaseName,
+		] = $array;
+		$this->port = (int)$port;
+	}
+
+	public static function determineConfigEnvironmentFile(): string {
```

Comment:
Are these two methods public solely for the purpose of accessing them in unit testing? If so, I still feel that the interface should reflect its intended use, but could be convinced otherwise. It looks like there is a way to unit test private methods without making them public -- what do you think about https://jtreminio.com/blog/unit-testing-tutorial-part-iii-testing-protected-private-methods-coverage-reports-and-crap/#targeting-private-protected-methods-directly?

Reply:
tl;dr - Testing private and protected methods is an anti-pattern.

I feel like using reflection to circumvent access modifiers is very hacky, and that private / protected method business logic should be tested from the top-down (through the public interface accessing those methods).

However, in this case, there is no way to test that business logic from the top down. Lets say we put all of the contents in that constructor, how can you test that each branch is being hit for explicitly determining the file name for the environment? Also, the config validation requires an instance of the `Dotenv` backend, so we would have to basically pass that in as a dependency into the constructor in order to validate it.

IMHO, these methods are fine to be public; they don't mutate any state. Unfortunately, PHP does not have something like "package-private" like JAVA. If that were the case, I could just remove them from the public API, and as long as my test was in the same package I could test them that way.

Reply:
For this codebase, I think I would prefer visibility to indicate intended use. Otherwise, it sounds like the goal would be to make every private method public -- _for testing_ -- and that doesn't feel right to me. If we really want to test it outside the public API, then there is still a way to do it, and I'd rather the hacks be in testing infrastructure than to compromise the API.

I'm definitely open to discussing it further!

Reply:
I'm happy to leave these as public, as you say they don't mutate any state. We're obviously coming at this from different angles, but I'm willing to see where this leads. :)

Reply:
I can inject the `Dotenv` dependency in the constructor and use a spy in order to verify interactions of the object. The dependency injection container stuff is in the next phase of my refactoring, so can I just do it there?

Reply:
Sure, go for it! It sounds like your plans to use DI here will be the best of both worlds.

Since I may not have been clear about why I commented on the method visibility in the first place, let me add some context. (I'm not advocating or lobbying for any change here, just hoping to provide some additional solid ground for any future discussions on the topic. This may all be obvious to you, but I don't think it hurts to be explicit!)

Let's say I'm a new developer to the project, and I decide to work on the `MySqlDatabase` class. I see that I need to use a `MySqlProperties` object. I head over to https://smrealms.github.io/api-docs and look at the public interface, and see some straightforward getter methods, but also some methods like `validateConfig`. Since it's public, it would be reasonable to ask "Do I need to call that? If I don't, does the config not get validated?". Instead of the API being prescriptive, I have to look at the source code to check. I'll then see that it's already called in the constructor, so I can infer that I don't need to call it myself. But that's more work, and could become burdensome if the entire codebase is written that way.

I see (at least) 3 possible ways to resolve such an issue to make it easier for a developer to use one part of the code without needing to know the internals the classes it uses:

1. Make the method private (though this can impede unit testing).
2. Document the method so that it's clear if/when someone would need to call it (not prescriptive, but better than nothing!).
3. Redesign the class to avoid the issue in the first place (e.g. your DI efforts).